### PR TITLE
POST /api/articles/:article_id/comments tested and complete

### PR DESCRIPTION
**Pull request by DevDolphin7**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -275,6 +275,75 @@ describe("/api/articles/:article_id/comments", () => {
...
+
+            const validKeys = [
+              "comment_id",
+              "votes",
+              "created_at",
+              "author",
+              "body",
+              "article_id",
+            ];
+            for (let i = 0; i < keys.length; i++) {
```
Comment (nc-niamh-smith) [inline __tests__/app.test.js:302]: a length check should happen before this, as it's dependent on the length of the keys array :)

File: __tests__/app.test.js [hunk 2]
```diff
@@ -275,6 +275,75 @@ describe("/api/articles/:article_id/comments", () => {
...
+          });
+      });
+
+      test("400: Responds with 'Bad request' where the posted data is not in format { username: `<string>`, body: `<string>` }", () => {
+        return request(app)
+          .post("/api/articles/7/comments")
+          .send({
+            username: "rogersop",
+            body: "Hello, world!",
+            sneakyKey: "drop all databases?",
```
Comment (nc-niamh-smith) [inline __tests__/app.test.js:338]: love the sneaky key

File: __tests__/app.test.js [hunk 3]
```diff
@@ -275,6 +275,75 @@ describe("/api/articles/:article_id/comments", () => {
...
+          })
+          .expect(400)
+          .then(({ body }) => {
+            expect(body.message).toBe("Bad request");
+          });
+      });
+    });
+  });
```
Comment (nc-niamh-smith) [inline __tests__/app.test.js:346]: what if the username doesn't exist on the database? 
Or there is a missing key?

edit: IGNORE this you handle it in your utils :)

File: __tests__/model-utils.test.js [hunk 4]
```diff
@@ -30,3 +34,82 @@ describe("checkForeignPrimaryKey", () => {
...
+  });
+
+  test("Does not mutate the input", () => {
+    const input = { hello: "world", timeIs: 14.46, dataType: ["array", true] };
+    const sortOrder = ["hello", "timeIs", "dataType"];
+    const inputCopy = JSON.parse(JSON.stringify(input));
+    formatObjectToArray(input, sortOrder);
+    expect(input).toEqual(inputCopy);
+  });
+});
```
Comment (nc-niamh-smith) [inline __tests__/model-utils.test.js:115]: Really thorough testing, and you've stuck really stringently to the single purpose rule 🎉

**Review by nc-niamh-smith — COMMENTED** [review]
🎉 Lovely work here Dan, I've suggested some tests and then you solved it in your utils so ignore that 😸 Really well thought out work here, well done.

File: __tests__/app.test.js (hunk 3 above)
Comment (DevDolphin7) [reply __tests__/app.test.js:346]: I think utils covers missing key, I don't think it checks for a valid username. I assume that'd be another db.query(), will have a look at some point!

File: __tests__/model-utils.test.js (hunk 4 above)
Comment (DevDolphin7) [reply __tests__/model-utils.test.js:115]: Thanks Niamh!

**Pull request merged**